diff -Naur speexdsp-1.2rc3/speexdsp.pc.in speexdsp-1.2rc3.tpg/speexdsp.pc.in
--- speexdsp-1.2rc3/speexdsp.pc.in	2015-01-03 20:26:24.000000000 +0000
+++ speexdsp-1.2rc3.tpg/speexdsp.pc.in	2015-10-09 00:15:37.996470109 +0000
@@ -10,6 +10,6 @@
 Version: @PACKAGE_VERSION@
 Requires: @FFT_PKGCONFIG@
 Conflicts:
-Libs: -L${libdir} -lspeexdsp
+Libs: -lspeexdsp
 Libs.private: @LIBM@
 Cflags: -I${includedir}
